drivers/stepper/ti/drv8424.c

```diff
@@ -0,0 +1,680 @@
+
+static void drv8424_user_callback_work_fn(struct k_work *work)
+{
+	int ret;
+
+	/* Get containing data struct */
+	struct drv8424_data *driver_data =
+		CONTAINER_OF(work, struct drv8424_data, event_callback_work);
+
+	/* Get next pending event (if any) from message queue */
+	enum stepper_event event;
+
+	ret = k_msgq_get(&driver_data->event_msgq, &event, K_NO_WAIT);
+	if (ret != 0) {
+		/* No pending events */
+		return;
+	}
+
+	/* Run the callback */
+	if (driver_data->event_callback != NULL) {
+		driver_data->event_callback(driver_data->dev, event,
+					    driver_data->event_callback_user_data);
+	}
+
+	/* If there are more pending events, resubmit this work item to handle them */
+	if (k_msgq_num_used_get(&driver_data->event_msgq) > 0) {
+		k_work_submit(work);
+	}
+}
+
+static int drv8424_schedule_user_callback(struct drv8424_data *data, enum stepper_event event)
+{
+	int ret;
+
+	/* Place event in message queue to be picked up by work item */
+	ret = k_msgq_put(&data->event_msgq, &event, K_NO_WAIT);
+	if (ret != 0) {
+		LOG_WRN("%s: Too many concurrent events, dropping event of type %d",
+			data->dev->name, event);
+		return -ENOBUFS;
+	}
+
+	/* Submit work item */
+	ret = k_work_submit(&data->event_callback_work);
+	if (ret < 0) {
+		LOG_ERR("%s: Failed to submit work item (error %d)", data->dev->name, ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static void drv8424_positioning_top_interrupt(const struct device *dev, void *user_data)
+{
+	struct drv8424_data *data = (struct drv8424_data *)user_data;
+	const struct drv8424_config *config = (struct drv8424_config *)data->dev->config;
+
+	if (!data->constant_velocity && data->reference_position == data->target_position) {
+		/* Check if target position is reached */
+		counter_stop(config->counter);
+		if (data->event_callback != NULL) {
+			/* Ignore return value since we can't do anything about it anyway */
+			drv8424_schedule_user_callback(data, STEPPER_EVENT_STEPS_COMPLETED);
+		}
+		data->is_moving = false;
+		return;
+	}
+
+	/* Determine direction we're going in for counting purposes */
+	enum stepper_direction direction = 0;
+
+	if (data->constant_velocity) {
+		direction = data->constant_velocity_direction;
+	} else {
+		direction = (data->target_position >= data->reference_position)
+				    ? STEPPER_DIRECTION_POSITIVE
+				    : STEPPER_DIRECTION_NEGATIVE;
+	}
+
+	/* Switch step pin on or off depending position in step period */
+	if (data->step_signal_high) {
+		/* Generate a falling edge and count a completed step */
+		gpio_pin_set_dt(&config->step_pin, 0);
+		if (direction == STEPPER_DIRECTION_POSITIVE) {
+			data->reference_position++;
+		} else {
+			data->reference_position--;
+		}
+	} else {
+		/* Generate a rising edge */
+		gpio_pin_set_dt(&config->step_pin, 1);
+	}
+
+	data->step_signal_high = !data->step_signal_high;
+}
+
+static int drv8424_set_microstep_pin(const struct device *dev, const struct gpio_dt_spec *pin,
+				     int value, const char *pin_name)
```

Comment:
I wouldn't bother with `pin_name`, if those gpio calls fail the whole system is messed up real bad and knowing which pin failed does not really give you much, I would drop this, save few bytes of flash and ram